packages/dataviews/src/dataviews-layouts/grid/preview-size-picker.tsx

```diff
@@ -39,31 +45,41 @@ function useViewPortBreakpoint() {
 	return null;
 }
 
-export default function DensityPicker( {
-	density,
-	setDensity,
-}: {
-	density: number;
-	setDensity: React.Dispatch< React.SetStateAction< number > >;
-} ) {
+export function useChangeGridColumnsOnViewportChange() {
 	const viewport = useViewPortBreakpoint();
+	const context = useContext( DataViewsContext );
+	const view = context.view as ViewGrid;
 	useEffect( () => {
-		setDensity( ( _density ) => {
-			if ( ! viewport || ! _density ) {
-				return 0;
-			}
-			const breakValues = viewportBreaks[ viewport ];
-			if ( _density < breakValues.min ) {
-				return breakValues.min;
-			}
-			if ( _density > breakValues.max ) {
-				return breakValues.max;
-			}
-			return _density;
-		} );
-	}, [ setDensity, viewport ] );
+		const gridColumns = view.layout?.gridColumns;
+		let newGridColumns;
+		if ( ! viewport || ! gridColumns ) {
+			return;
+		}
+		const breakValues = viewportBreaks[ viewport ];
+		if ( gridColumns < breakValues.min ) {
+			newGridColumns = breakValues.min;
+		}
+		if ( gridColumns > breakValues.max ) {
+			newGridColumns = breakValues.max;
+		}
+		if ( newGridColumns ) {
+			context.onChangeView( {
+				...view,
+				layout: {
+					...view.layout,
+					gridColumns: newGridColumns,
+				},
+			} );
+		}
```

Comment:
I'm not sure we should be changing the "view" object when the viewport changes. I feel the layout should be smart enough to render things properly without the need for the "view" object to change.

Reply:
This updates the view only if we had set the `previewSize` before and the new viewport (after some change) would have set a different value **if** a user would start fresh with the same viewport and updated the preview size picker value at the same value as the first action. Make sense?

Check the [video here ](https://github.com/WordPress/gutenberg/pull/67170#discussion_r1851562772)for when it actually updates.

Reply:
For me, the user is setting a "preview size", so we shouldn't change it dynamically for him. We should just make sure the Grid layout renders things properly by taking into account the "preview size" set by the user in the view object and the current viewport. 

Think about custom views for instance, it doesn't make sense to make a custom view "dirty" just because you resize the page or changed the orientation of your phone.

Reply:
I know it's a bit weird and had discussed the same thing with Jorge, when he first implemented that. I don't have strong opinions on keeping it, but the issue stems from the fact that we change dynamically (based on viewport) both available options for the size control and the max/min columns per size option. So what this hook tries to do is much these dynamic changes.

So the concept is: if a user selects the max size in a viewport (where we have 2 available options and displays 1 column) if they resize to very big viewport the hook adjusts to the max size again (where the available size options are 4 and display 3 columns). In essence it makes the view 'dirty' to try to preserve the original user size selection.

I feel like I can't explain it very well.. Or you fully get it and you just don't think we should do it?

Reply:
I understand very well what's happening, what I'm saying is that I don't think it's the right thing to do. WE should not be changing user choices just because he moved his phone or viewport. That doesn't mean the output shouldn't change, I'm saying the object "view" shouldn't change.

The dynamic computation need to be moved to the rendering phase (Grid component) and not be updated "view" object. 

If this was already present in trunk, I'm ok keeping it but this is a bug for me that should be fixed.

Reply:
Got it now. I updated [here](https://github.com/WordPress/gutenberg/pull/67170/commits/e41a211d83d102707e83b4d2f2e2207ecf5a2649).